aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -0,0 +1,896 @@
+
+  RNNDescriptors descs(fn, handle, x, y, hx, cx);
+
+  FilterDescriptor w_desc;
+  w_desc.set(weight_buf, 3);
+
+  size_t workspace_size;
+  auto x_descs_arr = descs.get_x_descs();
+  auto y_descs_arr = descs.get_y_descs();
+  CUDNN_CHECK(cudnnGetRNNWorkspaceSize(
+        handle,
+        descs.rnn_desc.desc(),
+        fn.tensors.seq_length,
+        x_descs_arr.data(),
+        &workspace_size
+        ));
+  Tensor workspace = input.type().toScalarType(kByte).tensor(workspace_size);
+
+  CUDNN_CHECK(cudnnRNNBackwardWeights(
+        handle,
+        descs.rnn_desc.desc(),
+        fn.tensors.seq_length,
+        x_descs_arr.data(), x.data_ptr(),
+        descs.hx_desc.desc(), hx.data_ptr(),
+        y_descs_arr.data(), y.data_ptr(),
+        workspace.data_ptr(), workspace.size(0),
+        w_desc.desc(), dw.data_ptr(),
+        fn_reserve.data_ptr(), fn_reserve.size(0)
+        ));
+
+  std::vector<Tensor> grad_weight_arr;
+  grad_weight_arr.reserve( weight.numel() );
+  for (const auto& w : weight_arr) {
+    grad_weight_arr.emplace_back(w.type().tensor(w.sizes()).zero_());
+  }
+
+  std::vector<Tensor> grad_params_arr;
+  size_t grad_params_stride0;
+  std::tie(grad_params_arr, grad_params_stride0) = get_parameters(handle, fn.rnn, descs.rnn_desc, descs.x_descs[0], w_desc, dw);
+  _copyParams(MatrixRef<Tensor>{grad_params_arr, grad_params_stride0},
+              MatrixRef<Tensor>{grad_weight_arr, static_cast<size_t>(weight_stride0)});
+
+  return grad_weight_arr; // stride is known from call site (and also inconvenient to return)
+}
+
+// We need this dispatcher because _cudnn_rnn_backward_weight has a stringent
+// ordering requirement with _cudnn_rnn_backward_grad
+std::tuple<Tensor, Tensor, Tensor, std::vector<Tensor>> _cudnn_rnn_backward(
+    const Tensor& input, TensorList weight, int64_t weight_stride0, const Tensor& weight_buf, const Tensor& hx, const Tensor& cx,
+    const Tensor& output, const Tensor& grad_output_r, const Tensor& grad_hy_r,
+    const Tensor& grad_cy_r,
+    int64_t mode, int64_t hidden_size,
+    int64_t num_layers, bool batch_first, double dropout,
+    bool train, bool bidirectional, IntList batch_sizes,
+    const Tensor& dropout_state, const Tensor& reserve,
+    std::array<bool, 4> output_mask
+    ) {
+
+  auto grad_output = grad_output_r.defined() ? grad_output_r : output.type().zeros_like(output);
+  auto grad_hy = grad_hy_r.defined() ? grad_hy_r : hx.type().zeros_like(hx);
+  auto grad_cy = cx.defined() ? (grad_cy_r.defined() ? grad_cy_r : cx.type().zeros_like(cx)) : grad_cy_r;
+
+  Tensor dx, dhx, dcx;
+  // NB: unconditionally compute this gradient, because it mutates reserve
+  std::tie(dx, dhx, dcx) = at::native::_cudnn_rnn_backward_grad(input, weight_buf, hx, cx, output, grad_output, grad_hy, grad_cy, mode, hidden_size, num_layers, batch_first, dropout, train, bidirectional, batch_sizes, dropout_state, reserve);
+  std::vector<Tensor> dw;
+  if (output_mask[3]) {
```

Comment:
Is it still `output_mask[3]` even when you use a cell that doesn't have `cx`?

Reply:
Yes, you always allocate an `output_mask` for every entry even when it is undefined.

Evidence:

```
  if (should_compute_output({ input_ix, hx_ix, cx_ix, weight_ix })) {
      auto grad_input_mask = std::array<bool, 4>{
      should_compute_output({ input_ix }),
      should_compute_output({ hx_ix }),
      should_compute_output({ cx_ix }),
      should_compute_output({ weight_ix }),
    };
```